File: algopt/rebalancer/examples/website/solvers/move_types/kl_search_examples.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * KLSearch Move Type Reference Examples
 *
 * This file demonstrates all usage patterns shown in the reference
 * documentation. Each function is a complete, runnable example.
 * */

#include "algopt/rebalancer/interface/ProblemSolver.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/ProblemSpecs_types.h"
#include "algopt/rebalancer/interface/thrift/gen-cpp2/SolverSpecs_types.h"

#include <folly/executors/CPUThreadPoolExecutor.h>
#include <folly/init/Init.h>

#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace facebook::rebalancer::interface;

// quick_example_start
void quickExample() {
  // Enable KL Search for advanced optimization
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  LocalSearchSolverSpec solverSpec;

  KLSearchMoveTypeSpec klSearchSpec; // No parameters needed

  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().klSearchMoveTypeSpec() = klSearchSpec;

  solver.addSolver(solverSpec);
}
// quick_example_end

void quickExampleComplete() {
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");
  solver.setObjectName("task");
  solver.setContainerName("server");

  // Setup imbalanced problem that may have local optima
  solver.setAssignment(
      std::map<std::string, std::vector<std::string>>{
          {"server0", {"task0", "task1", "task2", "task3"}},
          {"server1", {"task4"}},
          {"server2", {}},
      });

  std::map<std::string, double> cpuDim = {
      {"task0", 1.5},
      {"task1", 2.0},
      {"task2", 1.5},
      {"task3", 1.0},
      {"task4", 4.0},
  };
  solver.addObjectDimension("cpu", std::move(cpuDim));

  BalanceSpec balanceSpec;
  balanceSpec.name() = "balance-cpu";
  balanceSpec.scope() = "server";
  balanceSpec.dimension() = "cpu";
  solver.addGoal(std::move(balanceSpec), 1.0);

  LocalSearchSolverSpec solverSpec;
  KLSearchMoveTypeSpec klSearchSpec;
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().klSearchMoveTypeSpec() = klSearchSpec;
  solver.addSolver(solverSpec);

  auto solution = solver.solve();
  std::cout << "  Initial objective: " << *solution.initialObjective()->value()
            << "\n";
  std::cout << "  Final objective: " << *solution.finalObjective()->value()
            << "\n";
  std::cout << "  Improvement: "
            << (*solution.initialObjective()->value() -
                *solution.finalObjective()->value())
            << "\n";
}

// basic_usage_start
void basicUsage() {
  // Use KL Search to escape local optima
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");
  solver.setObjectName("task");
  solver.setContainerName("server");

  LocalSearchSolverSpec solverSpec;

  KLSearchMoveTypeSpec klSearchSpec;

  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().klSearchMoveTypeSpec() = klSearchSpec;

  solver.addSolver(solverSpec);
}
// basic_usage_end

// combined_start
void combined() {
  // Start with fast moves, then use KL Search for refinement
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");

  LocalSearchSolverSpec solverSpec;

  // Fast initial optimization
  SingleMoveTypeSpec singleSpec;
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().singleMoveTypeSpec() = singleSpec;

  // Refine and escape local optima
  KLSearchMoveTypeSpec klSearchSpec;
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().klSearchMoveTypeSpec() = klSearchSpec;

  solver.addSolver(solverSpec);
}
// combined_end

// balancing_start
void balancing() {
  // Use KL Search when simple moves plateau during load balancing
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");
  solver.setObjectName("shard");
  solver.setContainerName("node");

  LocalSearchSolverSpec solverSpec;

  // Initial balancing
  SingleMoveTypeSpec singleSpec;
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().singleMoveTypeSpec() = singleSpec;

  // Escape local optima for better balance
  KLSearchMoveTypeSpec klSearchSpec;
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().klSearchMoveTypeSpec() = klSearchSpec;

  solver.addSolver(solverSpec);
}
// balancing_end

void balancingComplete() {
  auto executor = std::make_shared<folly::CPUThreadPoolExecutor>(4);
  ProblemSolver solver(executor, "example", "test");
  solver.setObjectName("shard");
  solver.setContainerName("node");

  // Setup problem with potential local optima
  solver.setAssignment(
      std::map<std::string, std::vector<std::string>>{
          {"node0", {"shard0", "shard1", "shard2"}},
          {"node1", {"shard3", "shard4"}},
          {"node2", {"shard5"}},
          {"node3", {}},
      });

  std::map<std::string, double> loadDim = {
      {"shard0", 2.0},
      {"shard1", 1.5},
      {"shard2", 2.5},
      {"shard3", 3.0},
      {"shard4", 2.0},
      {"shard5", 4.0},
  };
  solver.addObjectDimension("load", std::move(loadDim));

  BalanceSpec balanceSpec;
  balanceSpec.name() = "balance-load";
  balanceSpec.scope() = "node";
  balanceSpec.dimension() = "load";
  solver.addGoal(std::move(balanceSpec), 1.0);

  LocalSearchSolverSpec solverSpec;
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().singleMoveTypeSpec() = SingleMoveTypeSpec{};
  solverSpec.moveTypeList()->emplace_back();
  solverSpec.moveTypeList()->back().klSearchMoveTypeSpec() =
      KLSearchMoveTypeSpec{};
  solver.addSolver(solverSpec);

  auto solution = solver.solve();
  std::cout << "  Initial objective: " << *solution.initialObjective()->value()
            << "\n";
  std::cout << "  Final objective: " << *solution.finalObjective()->value()
            << "\n";
  std::cout << "  Improvement: "
            << (*solution.initialObjective()->value() -
                *solution.finalObjective()->value())
            << "\n";
}

int main(int argc, char* argv[]) {
  const folly::Init init(&argc, &argv);
  std::cout << "=== Quick example showing basic KLSearch usage ===\n\n";

  std::cout << "Running quick_example...\n";
  quickExample();
  std::cout << "✓ quick_example completed\n\n";

  return 0;
}

File: algopt/rebalancer/examples/website/solvers/move_types/kl_search_examples.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
KLSearch Move Type Reference Examples

This file demonstrates all usage patterns shown in the reference documentation.
Each function is a complete, runnable example.
"""

# pyre-strict
from algopt.rebalancer.interface.py_client.ProblemSolver import ProblemSolver
from rebalancer.interface.thrift.v2.ProblemSolver.thrift_types import (
    GoalSpecs,
    SolverSpecs,
)
from rebalancer.interface.thrift.v2.ProblemSpecs.thrift_types import BalanceSpec
from rebalancer.interface.thrift.v2.SolverSpecs.thrift_types import (
    KLSearchMoveTypeSpec,
    LocalSearchSolverSpec,
    SingleMoveTypeSpec,
)


def quick_example():
    """Quick example showing basic KLSearch usage."""
    solver = ProblemSolver(service_name="example", service_scope="test")
    solver.setObjectName("task")
    solver.setContainerName("server")

    # quick_example_start
    # Enable KL Search for advanced optimization
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                moveTypeList=[
                    KLSearchMoveTypeSpec(),  # No parameters needed
                ]
            )
        )
    )
    # quick_example_end

    # Setup imbalanced problem that may have local optima
    solver.setAssignment(
        {
            "server0": ["task0", "task1", "task2", "task3"],
            "server1": ["task4"],
            "server2": [],
        }
    )

    solver.addObjectDimension(
        "cpu",
        {
            "task0": 1.5,
            "task1": 2.0,
            "task2": 1.5,
            "task3": 1.0,
            "task4": 4.0,
        },
    )

    solver.addGoal(
        GoalSpecs(
            balanceSpec=BalanceSpec(name="balance-cpu", scope="server", dimension="cpu")
        ),
        weight=1.0,
    )

    # Solve and print results
    solution = solver.solve()
    print(f"  Initial objective: {solution.initialObjective:.4f}")
    print(f"  Final objective: {solution.finalObjective:.4f}")
    print(f"  Improvement: {solution.initialObjective - solution.finalObjective:.4f}")


def basic_usage():
    """Use KL Search to escape local optima."""
    # basic_usage_start
    # Use KL Search to escape local optima
    solver = ProblemSolver(service_name="example", service_scope="test")
    solver.setObjectName("task")
    solver.setContainerName("server")

    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                moveTypeList=[
                    KLSearchMoveTypeSpec(),
                ]
            )
        )
    )
    # basic_usage_end


def combined():
    """Start with fast moves, then use KL Search for refinement."""
    # combined_start
    # Start with fast moves, then use KL Search for refinement
    solver = ProblemSolver(service_name="example", service_scope="test")

    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                moveTypeList=[
                    SingleMoveTypeSpec(),  # Fast initial optimization
                    KLSearchMoveTypeSpec(),  # Refine and escape local optima
                ]
            )
        )
    )
    # combined_end


def balancing():
    """Escape local optima when balancing load."""
    solver = ProblemSolver(service_name="example", service_scope="test")
    solver.setObjectName("shard")
    solver.setContainerName("node")

    # balancing_start
    # Use KL Search when simple moves plateau during load balancing
    solver.addSolver(
        SolverSpecs(
            localSearchSolverSpec=LocalSearchSolverSpec(
                moveTypeList=[
                    SingleMoveTypeSpec(),  # Initial balancing
                    KLSearchMoveTypeSpec(),  # Escape local optima for better balance
                ]
            )
        )
    )
    # balancing_end

    # Setup problem with potential local optima
    solver.setAssignment(
        {
            "node0": ["shard0", "shard1", "shard2"],
            "node1": ["shard3", "shard4"],
            "node2": ["shard5"],
            "node3": [],
        }
    )

    solver.addObjectDimension(
        "load",
        {
            "shard0": 2.0,
            "shard1": 1.5,
            "shard2": 2.5,
            "shard3": 3.0,
            "shard4": 2.0,
            "shard5": 4.0,
        },
    )

    solver.addGoal(
        GoalSpecs(
            balanceSpec=BalanceSpec(name="balance-load", scope="node", dimension="load")
        ),
        weight=1.0,
    )

    # Solve and print results
    solution = solver.solve()
    print(f"  Initial objective: {solution.initialObjective:.4f}")
    print(f"  Final objective: {solution.finalObjective:.4f}")
    print(f"  Improvement: {solution.initialObjective - solution.finalObjective:.4f}")


def main():
    print("=== KLSearch Move Type Examples ===\n")

    print("Running quick_example...")
    quick_example()
    print("✓ quick_example completed\n")

    print("Running balancing...")
    balancing()
    print("✓ balancing completed\n")


if __name__ == "__main__":
    main()
